pkgs/io_file/lib/src/vm_windows_file_system.dart

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2025, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:ffi';
+import 'dart:io' as io;
+
+import 'package:ffi/ffi.dart';
+import 'package:win32/win32.dart' as win32;
+
+import 'file_system.dart';
+
+String _formatMessage(int errorCode) {
+  final buffer = win32.wsalloc(1024);
+  try {
+    final result = win32.FormatMessage(
+      win32.FORMAT_MESSAGE_FROM_SYSTEM | win32.FORMAT_MESSAGE_IGNORE_INSERTS,
+      nullptr,
+      errorCode,
+      0, // default language
+      buffer,
+      1024,
+      nullptr,
+    );
+    if (result == 0) {
+      return '';
+    } else {
+      return buffer.toDartString();
+    }
+  } finally {
+    win32.free(buffer);
+  }
+}
+
+Exception _getError(int errorCode, String message, String path) {
+  final osError = io.OSError(_formatMessage(errorCode), errorCode);
+
+  switch (errorCode) {
+    case win32.ERROR_ACCESS_DENIED:
+    case win32.ERROR_CURRENT_DIRECTORY:
+    case win32.ERROR_WRITE_PROTECT:
+    case win32.ERROR_BAD_LENGTH:
+    case win32.ERROR_SHARING_VIOLATION:
+    case win32.ERROR_LOCK_VIOLATION:
+    case win32.ERROR_NETWORK_ACCESS_DENIED:
+    case win32.ERROR_DRIVE_LOCKED:
+      return io.PathAccessException(path, osError, message);
+    case win32.ERROR_FILE_EXISTS:
+    case win32.ERROR_ALREADY_EXISTS:
+      return io.PathExistsException(path, osError, message);
+    case win32.ERROR_FILE_NOT_FOUND:
+    case win32.ERROR_PATH_NOT_FOUND:
+    case win32.ERROR_INVALID_DRIVE:
+    case win32.ERROR_INVALID_NAME:
+    case win32.ERROR_NO_MORE_FILES:
+    case win32.ERROR_BAD_NETPATH:
+    case win32.ERROR_BAD_NET_NAME:
+    case win32.ERROR_BAD_PATHNAME:
+      return io.PathNotFoundException(path, osError, message);
+    default:
+      return io.FileSystemException(message, path, osError);
+  }
+}
+
+/// A [FileSystem] implementation for Windows systems.
+base class WindowsFileSystem extends FileSystem {
+  @override
+  void rename(String oldPath, String newPath) {
+    // Calling `GetLastError` for the first time causes the `GetLastError`
+    // symbol to be loaded, which resets `GetLastError`. So make a harmless
+    // call before the value is needed.
+    win32.GetLastError();
+    if (win32.MoveFileEx(
+          oldPath.toNativeUtf16(),
+          newPath.toNativeUtf16(),
```

Comment:
The memory allocated for these two paths must be freed manually or by wrapping this function with [using](https://pub.dev/documentation/ffi/latest/ffi/using.html) and passing `arena` as the allocator argument in `toNativeUtf16`:

```dart
void rename(String oldPath, String newPath) {
  using((arena) {
    // ...
    if (win32.MoveFileEx(
          oldPath.toNativeUtf16(allocator: arena),
          newPath.toNativeUtf16(allocator: arena),
    // ...
```

Reply:
> Done, and great drive by code review!

Thanks!

> Do you want to review more windows file system code?

Sure! Feel free to send review requests in future PRs. I'd be happy to review them.